File: archLintRules/src/main/java/me/ycdev/android/arch/lint/base/InheritDetectorBase.java
package me.ycdev.android.arch.lint.base;

import com.android.tools.lint.client.api.JavaEvaluator;
import com.android.tools.lint.detector.api.Detector;
import com.android.tools.lint.detector.api.JavaContext;

import org.jetbrains.uast.UClass;
import org.jetbrains.uast.UElement;

import java.util.HashSet;
import java.util.List;

public abstract class InheritDetectorBase extends Detector implements Detector.UastScanner {
    private HashSet<String> mWrapperClasses;

    /** Constructs a new {@link InheritDetectorBase} check */
    public InheritDetectorBase() {
        mWrapperClasses = getWrapperClasses();
    }

    protected abstract HashSet<String> getWrapperClasses();

    @Override
    public abstract List<String> applicableSuperClasses();

    protected abstract void reportViolation(JavaContext context, UElement element);

    @Override
    public void visitClass(JavaContext context, UClass cls) {
        String className = cls.getQualifiedName();
        if (mWrapperClasses.contains(className)) {
            return; // ignore the wrapper classes
        }

        JavaEvaluator evaluator = context.getEvaluator();
        boolean found = false;
        for (String wrapperClass : mWrapperClasses) {
            if (evaluator.inheritsFrom(cls, wrapperClass, false)) {
                found = true;
                break;
            }
        }
        if (!found) {
            reportViolation(context, cls);
        }
    }
}

File: archLintRules/src/main/java/me/ycdev/android/arch/lint/base/InheritDetectorBase.kt
package me.ycdev.android.arch.lint.base

import com.android.tools.lint.detector.api.Detector
import com.android.tools.lint.detector.api.JavaContext
import org.jetbrains.uast.UClass
import org.jetbrains.uast.UElement
import java.util.HashSet

abstract class InheritDetectorBase : Detector(), Detector.UastScanner {
    protected abstract val applicableClasses: List<String>
    protected abstract val wrapperClasses: HashSet<String>
    protected abstract fun reportViolation(context: JavaContext, element: UElement)

    override fun applicableSuperClasses(): List<String>? = applicableClasses

    override fun visitClass(context: JavaContext, declaration: UClass) {
        val wrappers = wrapperClasses
        val className = declaration.qualifiedName
        if (wrappers.contains(className)) {
            return // ignore the wrapper classes
        }

        val evaluator = context.evaluator
        var found = false
        for (wrapperClass in wrappers) {
            if (evaluator.inheritsFrom(declaration, wrapperClass, false)) {
                found = true
                break
            }
        }
        if (!found) {
            reportViolation(context, declaration)
        }
    }
}
